presto-orc/src/main/java/com/facebook/presto/orc/writer/FlatMapColumnWriter.java

```diff
@@ -0,0 +1,442 @@
+import com.facebook.presto.orc.metadata.OrcType.OrcTypeKind;
+import com.facebook.presto.orc.metadata.RowGroupIndex;
+import com.facebook.presto.orc.metadata.Stream;
+import com.facebook.presto.orc.metadata.statistics.ColumnStatistics;
+import com.facebook.presto.orc.metadata.statistics.IntegerStatisticsBuilder;
+import com.facebook.presto.orc.metadata.statistics.StatisticsBuilder;
+import com.facebook.presto.orc.metadata.statistics.StringStatisticsBuilder;
+import com.facebook.presto.orc.protobuf.ByteString;
+import com.facebook.presto.orc.stream.PresentOutputStream;
+import com.facebook.presto.orc.stream.StreamDataOutput;
+import com.google.common.collect.ImmutableList;
+import com.google.common.collect.ImmutableMap;
+import io.airlift.slice.Slice;
+import it.unimi.dsi.fastutil.longs.Long2ObjectMap;
+import org.openjdk.jol.info.ClassLayout;
+
+import java.io.IOException;
+import java.util.ArrayList;
+import java.util.List;
+import java.util.Map;
+import java.util.Optional;
+import java.util.SortedMap;
+import java.util.TreeMap;
+import java.util.concurrent.atomic.AtomicLong;
+import java.util.function.BiConsumer;
+import java.util.function.Supplier;
+
+import static com.facebook.presto.common.block.ColumnarMap.toColumnarMap;
+import static com.facebook.presto.orc.OrcEncoding.DWRF;
+import static com.facebook.presto.orc.metadata.ColumnEncoding.ColumnEncodingKind.DIRECT;
+import static com.facebook.presto.orc.metadata.ColumnEncoding.ColumnEncodingKind.DWRF_MAP_FLAT;
+import static com.facebook.presto.orc.metadata.ColumnEncoding.DEFAULT_SEQUENCE_ID;
+import static com.facebook.presto.orc.metadata.CompressionKind.NONE;
+import static com.facebook.presto.orc.metadata.DwrfMetadataWriter.getBytesKeyInfo;
+import static com.facebook.presto.orc.metadata.DwrfMetadataWriter.getIntKeyInfo;
+import static com.google.common.base.Preconditions.checkArgument;
+import static com.google.common.base.Preconditions.checkState;
+import static java.lang.Math.toIntExact;
+import static java.util.stream.Collectors.toList;
+
+public class FlatMapColumnWriter
+        implements ColumnWriter
+{
+    private static final int INSTANCE_SIZE = ClassLayout.parseClass(FlatMapColumnWriter.class).instanceSize();
+
+    private final int column;
+    private final int keyColumnId;
+    private final int valueColumnId;
+    private final MapType type;
+    private final boolean compressed;
+    private final ColumnEncoding columnEncoding;
+    private final CompressedMetadataWriter metadataWriter;
+    private final PresentOutputStream presentStream;
+
+    /* Fields for capturing column statistics */
+    private int nonNullValueCount;
+    private StatisticsBuilder statisticsBuilder;
+    private final Supplier<StatisticsBuilder> statisticsBuilderSupplier;
+    private final List<ColumnStatistics> rowGroupColumnStatistics = new ArrayList<>();
+    private final List<ColumnStatistics> keyRowGroupColumnStatistics = new ArrayList<>();
```

Comment:
Th keyRowGroupcolumnStatistics should be moved to the class that handles the key.